docs/design/copy-tracking.md

```diff
@@ -0,0 +1,286 @@
+info and needs none, but resolving the contents of F becomes more complicated.
+We need to (1) identify the greatest command ancestor of A and B (D)
+(which we do anyway), and (2) invoke `get_copy_records()` on F for each of
+`D::A` and `D::B` to identify the 'real' source file id for each parent. If
+these are the same, then we can use that as the base for a better 3-way merge.
+Otherwise, we must treat it as an add+add conflict where the base is the empty
+file id.
+
+It is possible that F1 and F2 both came from a common source file G, but that
+these copies precede D. In such case, we will not produce as good of a merge
+resolution as we theoretically could, but (1) this seems extremely niche and
+unlikely, and (2) we cannot reasonably achieve this without implementing some
+analogue of Mercurial's linknodes concept, and it would be nice to avoid that
+additional complexity.
+
+#### Squash conflicts
+
+Suppose we create commit A by renaming file F1 -> F, then we create child
+commit B in which we replace F by renaming F2 -> F. This touches on two issues.
+
+Firstly, if these two commits are squashed together, then we have a destination
+F with two copy sources, F1 and F2. In this case, we can store a
+`CopySources::Conflict([F1, F2])` as the copy source for F, and treat this
+commit as 'conflicted' in `jj log`. `jj status` will need modification to show
+this conflicted state, and `jj resolve` will need some way of handling the
+conflicted copy sources (possibly printing them in some structured text form,
+and using the user's merge tool to resolve them). Alternatively, the user can
+'resolve directly' by running `jj cp --after --resolve` with the desired copy
+info.
+
+Secondly, who is to say that commit B is 'replacing' F at all? In some version
+control systems, it is possible to 'integrate' a file X into an existing file Y,
+by e.g. propagating changes in X since its previous 'integrate' into Y, without
+erasing Y's prior history in that moment for the purpose of archaeology. With
+the commit metadata currently defined, it is not possible to distinguish
+between a 'replacement' operation and an 'integrate' operation.
+
+##### Track replacements explicitly
+
+One solution is to add a `destructive: bool` field or similar to the
+`CopySource` struct, to explicitly distinguish between these two types of copy
+records. It then becomes possible to record a non-destructive copy using
+`--after` to recognize that a file F was 'merged into' its destination, which
+can be useful in handling parallel edits of F that later sync this information.
+
+##### Always assume replacement
+
+Alternatively, we can keep copy-tracking simple in jj by taking a stronger
+stance here and treating all copies-onto-existing-files as 'replacement'
+operations. This makes integrations with more complex VCSs that do support
+'integrate'-style operations trickier, but it is possible that a more generic
+commit extension system is better suited to such backends.
```

Comment:
Ah, this is what I was asking about on the `CopySource::commit` field. Thanks for mentioning it.